packages/@webex/plugin-cc/src/services/core/WebSocket/connection-service.ts

```diff
@@ -0,0 +1,144 @@
+  private isKeepAlive: boolean;
+  private reconnectInterval: ReturnType<typeof setInterval>;
+  private webSocketManager: WebSocketManager;
+  private subscribeRequest: SubscribeRequest;
+
+  constructor(webSocketManager: WebSocketManager, subscribeRequest: SubscribeRequest) {
+    super();
+    this.webSocketManager = webSocketManager;
+    this.subscribeRequest = subscribeRequest;
+
+    this.isConnectionLost = false;
+    this.isRestoreFailed = false;
+    this.isSocketReconnected = false;
+    this.isKeepAlive = false;
+
+    this.webSocketManager.addEventListener('message', this.onPing);
+    this.webSocketManager.addEventListener('socketClose', this.onSocketClose);
+  }
+
+  private dispatchConnectionEvent(socketReconnected = false): void {
+    const event = new CustomEvent<ConnectionLostDetails>('connectionLost', {
+      detail: {
+        isConnectionLost: this.isConnectionLost,
+        isRestoreFailed: this.isRestoreFailed,
+        isSocketReconnected:
+          !this.webSocketManager.isSocketClosed && (socketReconnected || this.isSocketReconnected),
+        isKeepAlive: this.isKeepAlive,
+      },
+    });
+    this.dispatchEvent(event);
+  }
+
+  private handleConnectionLost = (): void => {
+    this.isConnectionLost = true;
+    this.dispatchConnectionEvent();
+  };
+
+  private clearTimerOnRestoreFailed = async () => {
+    if (this.reconnectInterval) {
+      clearInterval(this.reconnectInterval);
+    }
+  };
+
+  private handleRestoreFailed = async () => {
+    this.isRestoreFailed = true;
+    this.webSocketManager.shouldReconnect = false;
+    this.dispatchConnectionEvent();
+    await this.clearTimerOnRestoreFailed();
+  };
+
+  private updateConnectionData = (): void => {
+    this.isRestoreFailed = false;
+    this.isConnectionLost = false;
+    this.isSocketReconnected = false;
+  };
+
+  public setConnectionProp(prop: ConnectionProp): void {
+    this.connectionProp = prop;
+  }
+
+  private onPing = (event: Event): void => {
+    const msg = (event as CustomEvent<string>).detail;
+    const parsedEvent = JSON.parse(msg);
+    if (this.reconnectingTimer) {
+      clearTimeout(this.reconnectingTimer);
+    }
+    if (this.restoreTimer) {
+      clearTimeout(this.restoreTimer);
+    }
+    this.isKeepAlive = parsedEvent.keepalive === 'true';
+    const shouldUpdateConnectionData =
+      this.isKeepAlive || (this.isConnectionLost && !this.isRestoreFailed);
+    const shouldDispatchEvent =
+      this.isKeepAlive || (this.isConnectionLost && !this.isRestoreFailed);
+    const shouldDispatchEventWithReconnect = this.isSocketReconnected && this.isKeepAlive;
+
+    if (shouldUpdateConnectionData) {
+      this.updateConnectionData();
+    }
+
+    if (shouldDispatchEvent) {
+      this.dispatchConnectionEvent();
+    } else if (shouldDispatchEventWithReconnect) {
+      this.dispatchConnectionEvent(true);
+    }
+
+    this.reconnectingTimer = setTimeout(this.handleConnectionLost, this.wsDisconnectAllowed);
```

Comment:
We have a timer of 8 seconds, which closes the web socket if keep alive is false for 8 seconds (health check gets browser online status).
In that case, when keep alive is true once again, it restores the timer by simply re-registering the web socket connection (mentioned in handleSocketClose).
Logic is the same as agent desktop here.